### PR TITLE
EVG-20673 Allow expansions.update to set expansions to the same value as function vars

### PR DESCRIPTION
**Pull request by malikchaya2**
EVG-20673

### Description
Function vars wouldn't unset any vars whose values were changed by expansions.update. However, if it was set to the same value as the function var, it didn't work. We don't want that to be the case because we want users to be able to do this to persist function vars. 
image
 

### Testing
Unit test and tested on staging.

### CODE REVIEW

File: agent/command.go [hunk 1]
```diff
@@ -165,12 +165,20 @@ func (a *Agent) runCommand(ctx context.Context, tc *taskContext, logger client.L
...
-			for key, functionValue := range commandInfo.Vars {
-				currentValue := tc.taskConfig.Expansions.Get(key)
-				if currentValue != functionValue {
-					// If a command in the func updates the expansion value, don't reset it.
-					delete(prevExp, key)
+			// This defer ensures that the function vars do not persist in the expansions after the function is over
+			// unless they were updated using expansions.update
+
+			if cmd.Name() == "expansions.update" {
+				c, ok := cmd.(*command.Update)
```
Comment (Kimchelly) [inline agent/command.go:172]: I don't think this will work if the expansions are specified via a YAML file rather than explicitly listing them in `updates`. This is also a little janky because we're closely depending on the exact way that `expansions.update` is implemented and trying to reproduce its behavior here again.

Instead of doing it like this, can we consider an alternative implementation? One idea is adding a `TaskConfig.DynamicExpansions` field and modifying `expansions.update` to duplicate its writes there (and then unsetting the dynamic expansions when `runCommand` returns). Then, so long as `expansions.update` replicates its updates to the dynamic expansions, it gives a way for this defer to read the updated expansions, without relying on how exactly `expansions.update` is implemented or awkwardly type-casting.

File: agent/command_test.go [hunk 2]
```diff
@@ -328,20 +328,29 @@ functions:
...
 
 	key1Value := s.tc.taskConfig.Expansions.Get("key1")
-	s.Equal("expansion-update-value", key1Value, "key1 should be set to what it was updated to with expansions.update")
+	s.Equal("functionVar1", key1Value, "key1 should be set to what it was updated to with expansions.update")
+
+	key2Value := s.tc.taskConfig.Expansions.Get("key2")
+	s.Equal("functionVar2", key2Value, "key2 should be set to what it was updated to with expansions.update")
+
+	key3Value := s.tc.taskConfig.Expansions.Get("key3")
+	s.Equal("expansionVar3", key3Value, "key3 should be the original expansion value")
```
Comment (Kimchelly) [inline agent/command_test.go:354]: I would also add a test for expansions.update from YAML files since it missed that case.

File: agent/command_test.go [hunk 3]
```diff
@@ -328,20 +328,29 @@ functions:
...
+
```
Comment (Kimchelly) [inline agent/command_test.go:355]: nit: remove random whitespace.

**Review by Kimchelly — CHANGES_REQUESTED** [review]

File: agent/command/expansion_update.go [hunk 4]
```diff
@@ -73,6 +73,7 @@ func (c *update) ExecuteUpdates(ctx context.Context, conf *internal.TaskConfig)
 				return errors.WithStack(err)
 			}
 			conf.Expansions.Put(update.Key, newValue)
+			conf.DynamicExpansions.Put(update.Key, newValue)
```
Comment (Kimchelly) [inline agent/command/expansion_update.go:76]: Since `expansions.update` is the only usage of `DynamicExpansions`, can we have the command initialize `DynamicExpansions` in the case that it's nil? This also protects the agent from potential panics, where it assumes the `DynamicExpansions` exist.

Alternatively, explicitly testing a code path that runs `NewTaskConfig` and asserting that it must return a non-nil `DynamicExpansions` after initialization may be a reasonable substitute.

File: agent/command_test.go [hunk 5]
```diff
@@ -327,21 +328,56 @@ functions:
...
-	s.Equal("expansion-update-value", key1Value, "key1 should be set to what it was updated to with expansions.update")
+	key1Value := s.tc.taskConfig.Expansions.Get("key_1")
+	s.Equal("functionVar1", key1Value, "key_1 should be set to what it was updated to with expansions.update")
+
+	key2Value := s.tc.taskConfig.Expansions.Get("key_2")
+	s.Equal("functionVar2", key2Value, "key_2 should be set to what it was updated to with expansions.update")
+
+	key3Value := s.tc.taskConfig.Expansions.Get("key_3")
+	s.Equal("expansionVar3", key3Value, "key_3 should be the original expansion value")
+}
```
[the remark this answers is not in the document]
Comment (Kimchelly) [reply agent/command_test.go:356]: Can we also check that the dynamic expansions are reset to empty after the expansions.update command ends? Would avoid a potential bug of a kind similar to the original func vars bug if a dev changes the code in the future and forgets to reset the dynamic expansions (thus letting them carry over to later `expansions.update`s).

File: agent/command/expansion_update.go [hunk 6]
```diff
@@ -118,6 +120,10 @@ func (c *update) Execute(ctx context.Context,
 		if err != nil {
 			return errors.WithStack(err)
 		}
+		err = conf.DynamicExpansions.UpdateFromYaml(filename)
+		if err != nil {
```
Comment (Kimchelly) [inline agent/command/expansion_update.go:124]: nit: you can just combine L123-124 (same with L119-120).

File: agent/internal/task_config.go [hunk 7]
```diff
@@ -28,6 +28,7 @@ type TaskConfig struct {
 	Task               *task.Task
 	BuildVariant       *model.BuildVariant
 	Expansions         *util.Expansions
+	DynamicExpansions  *util.Expansions
```
Comment (Kimchelly) [inline agent/internal/task_config.go:31]: If you want this to be extra safe, having this be a `util.Expansions` without the pointer is better since you'll only have to deal with one level of nilness rather than two - for example, a nil `*util.Expansion` vs. a non-pointer `util.Expansions` that happens to be default initialize (i.e. a nil map).

File: agent/command/expansion_test.go [hunk 8]
```diff
@@ -38,7 +38,8 @@ func TestExpansionsPlugin(t *testing.T) {
 		expansions.Put("topping", "bacon")
 
 		taskConfig := internal.TaskConfig{
-			Expansions: &expansions,
+			Expansions:        &expansions,
+			DynamicExpansions: &util.Expansions{},
```
[the remark this answers is not in the document]
Comment (malikchaya2) [reply agent/command/expansion_test.go:42]: This is only used for unsetting vars and that is already tested extensively. Do you specifically want a test here?

File: agent/internal/task_config.go (hunk 7 above)
Comment (malikchaya2) [reply agent/internal/task_config.go:31]: Doesn't setting it in NewTaskConfig take care of this concern?

File: agent/command/expansion_update.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (malikchaya2) [reply agent/command/expansion_update.go:76]: I set it in NewTaskConfig for prod, and here I set it when setting up the test.

File: agent/command.go [hunk 9]
```diff
@@ -165,15 +165,19 @@ func (a *Agent) runCommand(ctx context.Context, tc *taskContext, logger client.L
...
 		if !tc.unsetFunctionVarsDisabled || tc.project.UnsetFunctionVars {
-			// This defer ensures that the function vars do not persist in the expansions after the function is over.
-			for key, functionValue := range commandInfo.Vars {
-				currentValue := tc.taskConfig.Expansions.Get(key)
-				if currentValue != functionValue {
-					// If a command in the func updates the expansion value, don't reset it.
-					delete(prevExp, key)
+			// This defer ensures that the function vars do not persist in the expansions after the function is over
+			// unless they were updated using expansions.update
+			if cmd.Name() == "expansions.update" && tc.taskConfig.DynamicExpansions != nil {
```
[the remark this answers is not in the document]
Comment (malikchaya2) [reply agent/command.go:170]: true, I'll add a length check. But I'd also rather keep both checks because there are no guarantees that DynamicExpansions won't be used in some other way in the future.

File: agent/command/expansion_update.go (hunk 4 above)
Comment (malikchaya2) [reply agent/command/expansion_update.go:76]: added to setup test

File: agent/command/expansion_test.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (malikchaya2) [reply agent/command/expansion_test.go:42]: I added a check here. But the test is not updated to test the file case. I don't think it's crucial to add considering the above.

File: agent/command.go [hunk 10]
```diff
@@ -165,15 +166,20 @@ func (a *Agent) runCommand(ctx context.Context, tc *taskContext, logger client.L
...
+				updatedExpansions := tc.taskConfig.DynamicExpansions.Map()
+				for k := range updatedExpansions {
+					if _, ok := commandInfo.Vars[k]; ok {
+						// If expansions.update updated this key, don't reset it
+						delete(prevExp, k)
+					}
 				}
 			}
 			tc.taskConfig.Expansions.Update(prevExp)
+			tc.taskConfig.DynamicExpansions = util.EmptyExpansion()
```
Comment (Kimchelly) [inline agent/command.go:181]: I'm pretty sure you can just call `util.NewExpansions(map[string]string{})` rather than make a brand new function for a specific case of it.

File: agent/command/expansion_update.go (hunk 4 above)
Comment (Kimchelly) [reply agent/command/expansion_update.go:76]: I still have this question, I meant adding something like this to the command's `Execute`, not just changing the test setup:
```
func Execute(...) {
// ...
if conf.DynamicExpansions == nil {
    // Initialize the map
}
// Actually do the expansions.update logic
}
```

File: agent/command_test.go [hunk 11]
```diff
@@ -328,20 +330,61 @@ functions:
...
 	key1Value := s.tc.taskConfig.Expansions.Get("key1")
-	s.Equal("expansion-update-value", key1Value, "key1 should be set to what it was updated to with expansions.update")
+	s.Equal("newValue1", key1Value, "key1 should be set to what it was updated to with expansions.update")
+
+	key2Value := s.tc.taskConfig.Expansions.Get("key2")
+	s.Equal("newValue2", key2Value, "key2 should be set to what it was updated to with expansions.update")
+
+	key3Value := s.tc.taskConfig.Expansions.Get("key3")
+	s.Equal("expansionVar3", key3Value, "key3 should be the original expansion value")
+	s.Equal(&util.Expansions{}, s.tc.taskConfig.DynamicExpansions)
```
Comment (Kimchelly) [inline agent/command_test.go:389]: nit: you can just check `s.Empty` rather than comparing it to the empty struct (same with the others).

File: agent/command_test.go [hunk 12]
```diff
@@ -82,6 +82,7 @@ func (s *CommandSuite) SetupTest() {
 		Identifier: "project_identifier",
 	}, &patch.Patch{}, util.Expansions{})
 	s.Require().NoError(err)
+	s.Equal(&util.Expansions{}, taskConfig.DynamicExpansions)
```
Comment (Kimchelly) [inline agent/command_test.go:85]: This isn't wrong, but it's a little weird to drive-by unit test the behavior of `internal.NewTaskConfig` in the agent suite. It seems like if we want to test the behavior of `internal.NewTaskConfig` itself, we should explicitly test that in the `internal` package, rather than drive-by test it in the setup here.

**Review by Kimchelly — CHANGES_REQUESTED** [review]

**Review by Kimchelly — APPROVED** [review]
I'm going to LGTM this a little prematurely so we can deploy it and continue with the function vars bugfix timeline. I read it over and it seems like it's overall a correct piece of code - we can address the other style/best practices comments in a follow-up PR.
